Vivek/banktransfer

vivekmenezes opened the pull request:
@petermattis I've added in a flag to disable the use of a transaction in the account transfers.

I also tried to change the readAllAccounts implementation to use a Scan() but it surprisingly didn't work. Doesn't the Scan() read the database at a timestamp?

petermattis commented on examples/bank/bank.go:
```diff
@@ -93,18 +96,84 @@ func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
+		// Copy responses into balances.
+		scan := call.Reply.(*proto.ScanResponse)
+		if len(scan.Rows) != bank.numAccounts {
+			log.Fatal(fmt.Errorf("Could only read %d of %d rows of the database.\n", len(scan.Rows), bank.numAccounts))
```
It's a bit strange to generate an error (`fmt.Errorf`) only to log it. Better to just use `log.Fatalf`.

vivekmenezes replied:
done

petermattis commented on examples/bank/bank.go:
```diff
@@ -139,7 +208,11 @@ func (bank *Bank) continuousMoneyTransfer() {
+			bank.moveMoneyTransaction(from, to, exchange)
+		} else {
+			bank.moveMoney(from, to, exchange)
+		}
```
A bit ugly to duplicate the functionality of `moveMoney` and `moveMoneyTransaction`. I think you can mostly unify the two and doing so will be even easier with the new client API.

petermattis commented on examples/bank/bank.go:
```diff
@@ -64,9 +67,9 @@ func readInt64(call client.Call) (int64, error) {
-// moveMoney() moves an amount between two accounts if the amount
-// is available in the from account. Returns true on success.
-func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
+// moveMoneyTransaction() moves an amount between two accounts in a transaction
```
godoc style is to start the comment with the name of the method (as you've done), but with the parentheses.

petermattis commented on examples/bank/bank.go:
```diff
@@ -93,18 +96,84 @@ func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
+	return true
+}
+
+// moveMoneyWithoutTransaction() moves an amount between two accounts without the
```
`s/moveMoneyWithoutTransaction()/moveMoney/g`

petermattis commented on examples/bank/bank.go:
```diff
@@ -93,18 +96,84 @@ func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
 }
 
 // Read the balances in all the accounts and return them.
+func (bank *Bank) readAllAccountsWithScan() []int64 {
```
What's not working here? The scan looks fine, especially within a transaction.

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -35,6 +36,8 @@ import (
 	"github.com/cockroachdb/cockroach/util/log"
 )
 
+var useTransaction = flag.Bool("use_transaction", true, "Turn off to disable transaction.")
```
doesn't really matter, but we've generally opted to use dashes (`use-transactions`).

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -64,9 +67,9 @@ func readInt64(call client.Call) (int64, error) {
+// if the amount is available in the from account. Returns true on success.
+func (bank *Bank) moveMoneyTransaction(from, to []byte, amount int64) bool {
 	// Early exit when from == to
 	if bytes.Compare(from, to) == 0 {
```
`bytes.Equal(from, to)`

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -93,18 +96,84 @@ func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
 		// Update both accounts.
-		txn.Prepare(client.Put(proto.Key(from), []byte(fmt.Sprintf("%d", fromValue-amount))))
-		txn.Prepare(client.Put(proto.Key(to), []byte(fmt.Sprintf("%d", toValue+amount))))
+		txn.Prepare(client.Put(proto.Key(from), []byte(fmt.Sprintf("%d", fromValue-amount))),
```
You don't need any of those `proto.Key` type conversions throughout the file. You can just pass `[]byte` in directly.

tschottdorf replied:
I think I asked before but I don't remember if I got an answer: You're deliberately not using Increment and our Integer types, right?

vivekmenezes replied:
On Sat, May 2, 2015 at 4:29 PM Tobias Schottdorf <notifications@github.com>
wrote:

> In examples/bank/bank.go
> <https://github.com/cockroachdb/cockroach/pull/881#discussion_r29550415>:
>
> > @@ -93,18 +96,84 @@ func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
> >  			return errRead
> >  		}
> >  		// Update both accounts.
> > -		txn.Prepare(client.Put(proto.Key(from), []byte(fmt.Sprintf("%d", fromValue-amount))))
> > -		txn.Prepare(client.Put(proto.Key(to), []byte(fmt.Sprintf("%d", toValue+amount))))
> > +		txn.Prepare(client.Put(proto.Key(from), []byte(fmt.Sprintf("%d", fromValue-amount))),
>
> I think I asked before but I don't remember if I got an answer: You're
> deliberately not using Increment and our Integer types, right?
>
Yeah I'm deliberately not using increment. I suppose we might want to hold
more than an amount in an account. Thanks for reviewing the PR.



> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/881/files#r29550415>.
>

tschottdorf commented on examples/bank/bank.go:
```diff
@@ -93,18 +96,84 @@ func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
+// Returns true on success.
+func (bank *Bank) moveMoney(from, to []byte, amount int64) bool {
+	// Early exit when from == to
+	if bytes.Compare(from, to) == 0 {
```
`bytes.Equal`

vivekmenezes commented:
@petermattis @tschottdorf ready for another viewing. This PR still uses the old API. I think it is valuable to check it in before transitioning this code to use the new API so that we can see the effect of the new API on the app in a different PR. Thanks

vivekmenezes committed:
Add a flag use_transaction to allow us to turn off transactions using
--use_transaction=false.
The program as expected fails almost immediately with --use_transaction off.

Also changed the scan of the database to be done using Scan().

Some minor changes.

vivekmenezes committed:
Add readAllAccountsWithScan().

vivekmenezes committed:
Consolidated the transaction and non-transaction transfers into a single method.

petermattis commented:
LGTM. Let's get this merged and start converting it to the new client API.
